File: C++/137. Single Number II.cpp
class Solution1 {
public:
    int singleNumber(vector<int>& nums) {
        int s1 = 0, s2 = 0, s3 = 0;
        for(auto num: nums)
        {
            s2 |= s1 & num;
            s1 ^= num;
            s3 = s1 & s2;
            s1 &= ~s3;
            s2 &= ~s3;
        }
        return s1;
    }
};

class Solution2 {
public:
    int singleNumber(vector<int>& nums) {
        int ans = 0;
        for(int i=0;i<32;i++)
        {
            int cnt = 0;
            for(auto num: nums)
            {
                if((num >> i) & 1 == 1)
                    cnt += 1;
            }
            cnt %= 3;
            if(cnt == 1)
                ans |= 1<<i;
        }
        return ans;
    }
};

File: Python/137. Single Number II.py
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ones, twos, threes = 0, 0 ,0 
        for num in nums:
            twos |= ones & num
            ones ^= num
            print(twos, ones)
            threes = ones & twos
            ones &= ~threes
            twos &= ~threes
            print(twos, ones)
        return ones